Source/AMRInterpolator/PunctureTracker.cpp

```diff
@@ -0,0 +1,239 @@
+    {
+        pout() << "Puncture " << ipuncture
+               << " restarted at : " << m_puncture_coords[ipuncture][0] << " "
+               << m_puncture_coords[ipuncture][1] << " "
+               << m_puncture_coords[ipuncture][2] << endl;
+        pout() << " with shift vector : " << m_puncture_shift[ipuncture][0]
+               << " " << m_puncture_shift[ipuncture][1] << " "
+               << m_puncture_shift[ipuncture][2] << endl;
+        pout() << "at time = " << a_current_time << endl;
+    }
+}
+
+//! Execute the tracking and write out
+void PunctureTracker::execute_tracking(double a_time, double a_restart_time,
+                                       double a_dt, const bool write_punctures)
+{
+    CH_TIME("PunctureTracker::execute_tracking");
+    if (m_num_punctures == 0)
+        return;
+    CH_assert(m_interpolator != nullptr); // sanity check
+
+    // get puncture coordinates and old shift value
+    std::vector<std::array<double, CH_SPACEDIM>> old_shift = m_puncture_shift;
+    CH_assert(m_puncture_coords.size() == m_num_punctures); // sanity check
+
+    // new shift value
+    interp_shift();
+
+    // update puncture locations using second order update
+    for (int ipuncture = 0; ipuncture < m_num_punctures; ipuncture++)
+    {
+        FOR1(i)
+        {
+            m_puncture_coords[ipuncture][i] +=
+                -0.5 * a_dt *
+                (m_puncture_shift[ipuncture][i] + old_shift[ipuncture][i]);
+        }
+    }
+
+    // print them out
+    if (write_punctures)
+    {
+        bool first_step = false;
+        SmallDataIO punctures_file(m_punctures_filename, a_dt, a_time,
+                                   a_restart_time, SmallDataIO::APPEND,
+                                   first_step);
+
+        // use a vector for the write out
+        punctures_file.write_time_data_line(get_puncture_vector());
+    }
+}
+
+//! Use the interpolator to get the value of the shift at
+//! given coords
+void PunctureTracker::interp_shift()
+{
+    CH_TIME("PunctureTracker::interp_shift");
+    // resize the vector to the number of punctures
+    m_puncture_shift.resize(m_num_punctures);
+
+    // refresh interpolator
+    m_interpolator->refresh();
+
+    // set up shift and coordinate holders
+    std::vector<double> interp_shift1(m_num_punctures);
+    std::vector<double> interp_shift2(m_num_punctures);
+    std::vector<double> interp_shift3(m_num_punctures);
+    std::vector<double> interp_x(m_num_punctures);
+    std::vector<double> interp_y(m_num_punctures);
+    std::vector<double> interp_z(m_num_punctures);
+
+    for (int ipuncture = 0; ipuncture < m_num_punctures; ipuncture++)
+    {
+        interp_x[ipuncture] = m_puncture_coords[ipuncture][0];
+        interp_y[ipuncture] = m_puncture_coords[ipuncture][1];
+        interp_z[ipuncture] = m_puncture_coords[ipuncture][2];
+    }
+
+    CH_assert(ChomboParameters::variable_name_to_enum("shift1") >= 0);
```

Comment:
Aha! So it is used outside of ChomboParameters @mirenradia :-) Looking forward to this moving to a refactored UserVariables!

Reply:
I'm not convinced this is what I had in mind when I wrote this. It's also an inefficiently written function (it creates a whole new array each time it's called) so I'm uncomfortable with its use here. Having said that, I can't really think of an alternative if this assert _must_ be here.

Reply:
Good to know @mirenradia. I'm not too worried if it is inefficient since it won't run on OPT=HIGH, but I'm also ok to remove the assert, because I think it is fine to assume that shift1 exists if you have a BBH. 

If we refactor the UserVariables so people can simply select an option to include CCZ4 vars as a block, then we could build in a function for includes_ccz4_vars() which could be checked instead.

Reply:
I think this (and below) can be changed back to just `c_shift1`, etc. now that this file has been moved to the new `BlackHoles` subdirectory which will only be included if you have a `c_shift1` in your `UserVariables.hpp` anyway.

I can now recover my claim that `variable_name_to_enum` won't be used outside of parameter loading :laughing: